[READY] Re-works the Durand ability

zxaber opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Replaces the old Durand defense ability with a new take on it. Adds a directional shield that blocks all attacks except explosives (and EMP damage) from the direction the Durand is facing. No longer restricts movement, but has a very heavy power cost.

As an aside, most of the Durand-related defense code has been moved from mecha_defense.dm to durand.dm. Also added in a return false for the mecha.dm's use_power() proc, which is used if the cell's proc returned false.

Sounds created by ActionNinja

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
The old Durand ability is pretty awful. Losing the ability to move (or turn) puts the ability into the "so situational it almost never comes up" category. Ideally, it would be something to toggle on and off in-between incoming shots, but that's not a great system when it's activated by an action button. Meanwhile, the effect _only_ being extra deflect chance isn't interesting or fun to play with.

This new version falls in-line with what the Durand should be good at (being a tanky weapons platform when attacking something head-on), without removing or alleviating the weaknesses of the Durand (easy to flank and circle). The shield does not block EMP effects or explosive damage, so core counters are not changed. The heavy power cost _really_ punishes people who just leave it on (it draws so much power that yellow cells will drain, not even counting the power cost for each shot blocked). In theory, this should make the Durand a viable alternative to the Gygax, and worth building even if Gygax mats are available.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: Zxaber
balance: Re-worked the Durand's defense ability into a directional shield that blocks all attacks except explosive and EMP-based. Does not inhibit movement or weapon use, but has a very high power cost.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

![image](https://user-images.githubusercontent.com/37497534/61105000-4a630e00-a42d-11e9-8e8b-d696f9a00885.png)
![durand_shield](https://user-images.githubusercontent.com/37497534/61528187-fcbd4700-a9d2-11e9-8038-11b3524eec59.gif)
https://streamable.com/a8w08

Power usage is as follows:
- Every two seconds, the Durand shield consumes 100 charge just for being on, as part of the process proc.
- When the shield is hit by an attack, it consumes an extra 100 charge for every point of damage it received.

The usage is interesting because it creates a neat dynamic with yellow core cells. While they regenerate power on their own, it's not instant, and they only have 10MJ of capacity (or 20 for hypercharged ones). They regen so slowly that just leaving the shield on will drain the power. A better cell like hypers or bluespace will let you last much longer in a fight, though you'll have to return to a proper charger at some point.

For comparison's sake, the Gygax ability sets the power cost per step from 10 to 300. 

--

The actual blocking of attacks works as follows.
![image](https://user-images.githubusercontent.com/37497534/61106489-31a92700-a432-11e9-942f-7190d18b7849.png)
If a melee attack comes from a mob standing in one of the red-shaded areas, or a projectile hits while in those areas, the attack is routed to the shield. Since projectiles travel at an angle directly influenced by where the attacker clicks, this opens up a neat mechanic where you can _just_ shoot around the shield if you make the bullet enter the tile to the side of the durand before it hits.

To-Do:
- [x] Sounds for enabling/disabling the shield, and for attacks being blocked by it
- [x] Possibly consider shield enable/disable animations, and also having it emit a tiny bit of light if the durand's light is not on.
- [x] AutoDoc compliance

zxaber committed:
Merge pull request #1 from tgstation/master

herp a derp this better not show up on the discord

zxaber committed:
Merge pull request #2 from tgstation/master

This better not pop up on the discord

zxaber committed:
Merge pull request #3 from tgstation/master

updoot

zxaber committed:
Merge pull request #4 from tgstation/master

Every time I do this I never know if I'm doing it right

zxaber committed:
Merge pull request #5 from tgstation/master

updoot

zxaber committed:
Woo shields

daklj commented:
buffing mechs lol

zxaber commented:
I _did_ just double-nerf them, to be fair.

PKPenguin321 commented:
no yeah this is a nerf for sure
not that that's necessarily a bad thing, at least it's more interesting than the old ability

Ratvar commented:
Full immunity to EMP at cost of using yellow slime cores? Mmmm.

TheChosenEvilOne commented:
> 
> 
> Full immunity to EMP at cost of using yellow slime cores? Mmmm.

Did you even read it? it clearly says "The shield does not block EMP effects or explosive damage, so core counters are not changed."

Ratvar commented:
Oh. I've read "Re-worked the Durand's defense ability into a directional shield that blocks non-explosive/EMP attacks." after that, so I guess *I did read*. Fix the changelog, I guess?

Whoneedspacee commented:
hey at least i might use the ability now and actually have a point to making a durand besides its cheap

zxaber commented:
It blocks attacks that _aren't_ EMP or explosives. Ergo, EMP attacks and explosions still hit the mech. Where did I mess up on the changelog?

FloranOtten commented:
This is still very situational. I can only forsee this being good against nukies and blobs. You'll rarely have the mats and the research for nukies, and blobs is a single trick pony. If this blocked not EMP damage but EMP weapons disable, at the cost of massive powerdrain, it would elevate the Durand to a level where it could at least be used when gygaxes are available.

zxaber commented:
My stance on that is I will put in the ability to block EMP effects on mechs the day I'm allowed to add in flash protection to borgs.

SgtxRootbeer commented:
i know balance would be thrown off by this, but perhaps maybe let it block explosions, at great cost.

it would be really, really cool to see an admin disable the bombcap, so some traitor dink can make a hyperbomb, and try to destroy the station, with only a shadow behind the hastily-activated durand left.

Alek2ander commented:
>it would be really, really cool to see an admin disable the bombcap, so some traitor dink can make a hyperbomb, and try to destroy the station, with only a shadow behind the hastily-activated durand left.

That, except the Durand should be wrecked by it. Great deeds require great sacrifice.

nemvar commented:
This is a big no from me until you actually move the defense ability and the defense mode var from mecha to durand. This just makes mech code even more cursed right now.

zxaber commented:
Most of the related code is now under the durand, with a small bit in the mecha actions.

nemvar commented on code/game/mecha/mecha_actions.dm:
```diff
@@ -164,10 +164,31 @@
+/**An object to take the hit for us when using the Durand's defense mode.
+It is spawned by the Durand (see the durand.dm file) when an attack is made and determined to be blocked by the shield.
+Attacks are passed to it after creation. It will deal the damage as power drain to the mech and then self-delete.*/
+/obj/mech_defence_mode_scapegoat //projectiles get passed to this when defence mode is enabled
```
This entire thing is pointless.

nemvar commented on code/game/mecha/mecha_actions.dm:
```diff
@@ -164,10 +164,31 @@
+		qdel(src)
+		return
+	. = ..()
+	if(!chassis.use_power((max_integrity - obj_integrity) * 100))
```
Less damage drains more power? Something is off with this calculation.

zxaber replied:
If I have 20 max health, and I lose 5, my health is left at 15. So 20 - 15 is 5 charge to lose.
If I have 20 max health, and I lose 7, my health is left at 14. So 20 - 14 is 7 charge to lose.
We then multiply it by 100 to get the actual charge used.

If chassis.use_power returns false, it's because the amount of power needed surpassed what was available. The cell proc that handles using power will just output a false in this situation, but doesn't drain the power at all. So instead we have to do it manually.

nemvar replied:
Yeah sorry, I forgot what a obj_integrity did. Not like this proc makes sense anyways. Should probably just be (. * 100) though.

Tlaltecuhtli replied:
what if i had 20 max health 10 health, lose 1, do i lose 1 power or 11?

zxaber replied:
The way this works is that a new entity is spawned in (with a shit load of health), damage is dealt to it, and then after calculating the charge to take, the entity is removed. The next time damage is done, a new entity is spawned in with a fresh health bar, to act as the scapegoat.

It's honestly probably not the sanest way to go about it, and I might change it. It was just a decent way in my mind of passing on damage and calculating from it without having to worry about how different damage types might affect things.

nemvar replied:
You can basically abstract this by moving this code to take_damage.
Something like this:
```
/obj/mecha/combat/durand/take_damage(damage_amount, damage_type = BRUTE, damage_flag = "", sound_effect = TRUE, attack_dir, armour_penetration = 0)
if(defence_check(damage_amount, attack_dir, damage_flag))
	return
. = ..()

/obj/mecha/combat/durand/proc/defence_check(damage, adir, damage_flag = "")
	if (!defence_mode)
		return
	if(damage_flag == "bomb")
		return
	switch(dir) //There's probably a better way, haven't checked.
		if (EAST)
			if(adir != WEST)
				return
		if (SOUTH)
			if(adir != NORTH)
				return
		if (WEST)
			if(adir != EAST)
				return
		if (NORTH)
			if(adir != SOUTH)
				return
	if(!use_power(damage * 100))
		defense_action.turnoff()
		return
	return TRUE

Tlaltecuhtli replied:
if(damage_flag == "stamina")
		return

nemvar replied:
That's nonsense. damage_flag is basically the type of armor you want to use. And objs already don't get damaged by anything that isn't BRUTE or BURN.

zxaber replied:
After toying with it for a bit, I'm not sure I want to re-do it like this.

The biggest issue is that diagonals don't neatly fall into a dir, so they're sorta fudged a bit. When you fire in a diagonal, the projectile is given a four-way-compatible dir, usually confined to one axis (I believe left and right). The result is that if the mech is facing two specific directions (I believe north or south), its field of defense will be a lot narrower than if it was facing the other directions. Ideally, I'd want the mech to perform consistently no matter what direction its facing.

That aside, melee attacks don't seem to give a dir, so I'd have to add extra checks in for that.

Shapsy commented:
We New Conglomerate now boys

fludd12 commented:
Do note that hypercharged slime cells also _regenerate_ power faster. Of course, not nearly fast enough to keep up with this, but still faster.

I'm eager to see a line of three durands marching in phalanx formation down a hall, utterly destroying all in their path until they get backstabbed by an assistant or tank-mine'd with a few IEDs.

Arathianen commented:
So, hot take of the day:

The ability should drain slightly _less_ than a hypercharged yellow core regenerates. I am....entirely unsure how much that is! But it's pretty fast.

My logic:

Gygaxes and durands are, in a vacuum, pretty equal. Durands have a massive hp and armour advantage vs speed for the gygax, but the gygax ability is so draining that can only be used for a few seconds and then the gygax can't do other stuff/has to go recharge.

And all of that is true!....if we only had t1 parts/batteries and so on. The thing is, no one is making a gygax or durand without t4 parts. In fact, making a durand without hypercharged yellow cores is a pretty big waste.

So the gygax beats the durand because, with the realistic loadout, it has an infinite - or very long - use of an extremely strong ability.

A normal durand (without hypercharged) would have to still turn the ability off the moment it doesn't need it (or at least limit it to potentially immediately dangerous situations) since it's still a slow ass mech that will take half a century to get back to the charging station. But a hypercharged durand can now have the ability turned on by default and only disabled when its battery goes dangerously slow from actual hits. That way the ability (and the durand) actually scale into late game properly.

Instead, what I would do to balance it, is up the power loss from direct hits. That way the shield can be relatively easier overwhelmed and forced to be turned off.

That way we can have durands that are, from the start, protected and the enemies know they gotta concentrate some force on the durand (or hit it from the back...or with ions...or blast) else it's protected. It makes the durand a proper "tank" and not just...well, something to hit and run with a toolbox until it either dies as it does now or it runs out of battery in a minute.

dee-dubya commented:
I agree entirely with Arathianen's post above. Especially considering we should also be encouraging more interdepartmental cooperation. If robotics actually is able to get their hands on a hypercharged yellow core, I think both mechs should be able to benefit equally from the boon.

CRITAWAKETS commented:
Maybe make the shield crack open if it takes too much damage within a small amount of time? A good mech user with a hypercharged core or even a regular yellow slime battery can basically just become invincible using this. I will teach you the secret to become invincible.


Just disable the shield when you arent being shot at lmao now you are invincible.

Also the ammo system was a buff to every single ammo-using weapon that was not the missiles by drastically increasing their clip size to ludicrous levels along with having ammo storage that is enough to kill a whole station and still have way more than you need.

Mechs need a nerf to literally everything.

zxaber commented:
The mech ammo change didn't affect the max ammo you can load into each weapon. It just added an ammo stock that weapons reload from, instead of using power to pull bullets from the aether.

shellspeed1 commented:
> 
> 
> So, hot take of the day:
> 
> The ability should drain slightly _less_ than a hypercharged yellow core regenerates. I am....entirely unsure how much that is! But it's pretty fast.
> 
> My logic:
> 
> Gygaxes and durands are, in a vacuum, pretty equal. Durands have a massive hp and armour advantage vs speed for the gygax, but the gygax ability is so draining that can only be used for a few seconds and then the gygax can't do other stuff/has to go recharge.
> 
> And all of that is true!....if we only had t1 parts/batteries and so on. The thing is, no one is making a gygax or durand without t4 parts. In fact, making a gygax without hypercharged yellow cores is a pretty big waste.
> 
> So the gygax beats the durand because, with the realistic loadout, it has an infinite - or very long - use of an extremely strong ability.
> 
> A normal durand (without hypercharged) would have to still turn the ability off the moment it doesn't need it (or at least limit it to potentially immediately dangerous situations) since it's still a slow ass mech that will take half a century to get back to the charging station. But a hypercharged durand can now have the ability turned on by default and only disabled when its battery goes dangerously slow from actual hits. That way the ability (and the durand) actually scale into late game properly.
> 
> Instead, what I would do to balance it, is up the power loss from direct hits. That way the shield can be relatively easier overwhelmed and forced to be turned off.
> 
> That way we can have durands that are, from the start, protected and the enemies know they gotta concentrate some force on the durand (or hit it from the back...or with ions...or blast) else it's protected. It makes the durand a proper "tank" and not just...well, something to hit and run with a toolbox until it either dies as it does now or it runs out of battery in a minute.

This is sadly not the case. In a fight between a gygax and a durand, the gygax will almost win. This is because the gygax is far more maneuverable, can punch through walls, and if it gets in melee range can just beat the shit out of a durand no problem. The only time this wouldn't happen is if you managed to discharge the battery of the gygax enough that it had to stop attacking.

nemvar commented:
If only the code improvements to mechs would be done with the same engagement as the discussion about it.

zxaber commented:
As it turns out, the ability already is ever so slightly less than the hypercharged yellow.

Of course, if you're taking a beating, 20MJ will only last you so long.

SkeletalElite commented:
A possible alternative to this that would make the durand worth using is making it so that EMPs dont disable its weapons whiles its in defense mode.

dee-dubya commented:
Here's a thought: Use the unstoppable force vs immovable object argument. Will a gygax in boost mode rapidly running into a Durand with the new shield ability both deplete a battery charge at roughly the same time?

Arathianen commented:
> As it turns out, the ability's drain already is ever so slightly less than the hypercharged yellow recharge rate.
> 
> Of course, if you're taking a beating, 20MJ will only last you so long.

Sounds good to me. It's how it should be then.

Looking forward to durands not being hot trash. They are my favourite type of mech.

zxaber commented:
@SkeletalElite
As I said earlier;

>My stance on that is I will put in the ability to block EMP effects on mechs the day I'm allowed to add in flash protection to borgs

YakumoChen commented:
Good change, Technically a buff but I would still rather see people with a reason to build durands than gygaxes

zxaber committed:
reshuffle, adding a sound

zxaber commented:
Right, so animations probably aren't happening. Byond likes to sync all animations up, and I can't figure out a feasible way to start an overlay from the beginning of its animation. If I ever figure a proper way (and/or if we get better control over it in the future), I can add it in later.

A sound effect for hitting the shield has been helpfully provided by ActionNinja. I might raise the volume on the file a bit, since it's a touch quiet when the sounds of gunfire are also around.

I've moved code from mecha_actions.dm to durand.dm, where it made sense to, and added a check so that thrown items are also blocked by the shield.

nemvar commented on code/game/mecha/mecha.dm:
```diff
@@ -110,7 +110,6 @@
 	//Action vars
 	var/thrusters_active = FALSE
 	var/defence_mode = FALSE
```
I meant this variable and the var in line 101. Probably out of scope of this PR because that would mean that you have to refactor mech UI and the simplemob mech pilot

zxaber replied:
Yeah, I tried to move defence_mode to durand.dm, but then mecha_topic.dm throws a fit about it.

zxaber committed:
Re-organizing

zxaber commented:
After getting a bit of help with things, I've re-done a lot of the PR. The Shield is now a permanent object (at least as permanent as the mech is), instead of being created and then destroyed every time it needs to take damage. This means I was able to get the shield deploy and drop animations to work correctly after all, and now it emits a nice blue light while active.

Using the power of S I G N A L S, _all_ of the activation shield code has been removed from mecha_actions.dm and placed into code under the shield object itself. Mecha_actions.dm simply has code for the action button, and the button itself only sends a signal to the mech (to relay to the shield) for activating/deactivating the shield.

I have been told by several sources that I should use signals for handling attacks being passed to the shield, so that's what I plan to work on next. I was getting runtimes when attempting it, but I only just found the cause, which was not even really related to using signals.

I'd also like the shield to not show up in context menus, so I need to figure that out too.

zxaber committed:
better defines for the signals I had

zxaber committed:
new stuff

zxaber commented:
More sounds from action, and an impact flicker.

Proper video, too:
https://streamable.com/a8w08

As seen in the video, emitters do a strange visual bug when they happen to be the final shot that causes the battery to hit zero and the shield to drop. Other projectiles don't do this, and neither do melee attacks, so I dunno what the deal is.

Projectiles are now blocked via signals. Other attacks don't seem as easy to implement signals for. If it's absolutely necessary for a merge, I can look into it.

zxaber committed:
fixes

zxaber committed:
Replaced all instances of "defence" with defense"

and other small changes.

zxaber committed:
oof

zxaber commented:
Alright, I _think_ I did autodoc things right. Far as I can tell, I'm ready for review.

zxaber committed:
spelling, and a forgotten check

actioninja commented on code/game/mecha/combat/durand.dm:
```diff
@@ -20,3 +34,178 @@
+		flick("shield_raise", src)
+		playsound(src, 'sound/mecha/mech_shield_raise.ogg', 50, FALSE)
+		set_light(l_range = MINIMUM_USEFUL_LIGHT_RANGE	, l_power = 5, l_color = "#00FFFF")
+		sleep(3)
```
avoid using sleep at all.
try setting icon state before the flick

zxaber replied:
The sleep command allows for a few fluff effects, like the shield not blocking damage until it's fully up and emitting light until the very last frame of the animation. It also handles setting the shield's invisibility when being dropped, so that disabled shields don't show up on the right-click menu.

I could probably replace it with timers and extra procs, but are sleeps really that bad of an issue?

actioninja replied:
I looked more into it, and this is fine.
I was under the impression that sleeps are always bad news, but they can be safe in some cases.

actioninja commented on code/game/mecha/combat/durand.dm:
```diff
@@ -20,3 +34,178 @@
+	obj_integrity = 10000
+
+/obj/durand_shield/play_attack_sound()
+	playsound(src, 'sound/mecha/mech_shield_deflect.ogg', 100, 1)
```
1 to TRUE

zxaber replied:
done

zxaber committed:
you, upgraded, improved, yes

zxaber commented on code/game/mecha/combat/durand.dm:
```diff
@@ -20,3 +34,178 @@
+/obj/mecha/combat/durand/proc/defense_check(var/turf/aloc)
+	if (!defense_mode || !shield || shield.switching)
+		return FALSE
+	var/blocked = FALSE
```
done

adamkad1 commented:
> My stance on that is I will put in the ability to block EMP effects on mechs the day I'm allowed to add in flash protection to borgs.

i want
